File: typescript/packages/core/src/core/slack/readdir.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { SlackAccessor } from '../../accessor/slack.ts'
import type { IndexEntry } from '../../cache/index/config.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { listChannels, listDms, type SlackChannel } from './channels.ts'
import { SlackIndexEntry } from './entry.ts'
import { fetchMessagesForDay, type SlackMessage } from './history.ts'
import { detectScope } from './scope.ts'
import { listUsers } from './users.ts'
import { stripSlash } from '../../util/slash.ts'

export const VIRTUAL_ROOTS = ['channels', 'dms', 'users'] as const

const SOFT_HISTORY_ERRORS = [
  'not_in_channel',
  'channel_not_found',
  'missing_scope',
  'is_archived',
  'not_authed',
]

function isSoftHistoryError(err: unknown): boolean {
  const msg = err instanceof Error ? err.message : String(err)
  return SOFT_HISTORY_ERRORS.some((code) => msg.includes(code))
}

export async function latestMessageTs(
  accessor: SlackAccessor,
  channelId: string,
): Promise<number | null> {
  let messages: { ts?: string }[]
  try {
    const data = await accessor.transport.call('conversations.history', {
      channel: channelId,
      limit: '1',
    })
    messages = (data.messages as { ts?: string }[] | undefined) ?? []
  } catch (err) {
    if (isSoftHistoryError(err)) return null
    throw err
  }
  if (messages.length === 0) return null
  return Number.parseFloat(messages[0]?.ts ?? '0')
}

export function dateRange(latestTs: number, created: number, maxDays = 90): string[] {
  const endMs = Math.floor(latestTs * 1000)
  const startMs = Math.floor(created * 1000)
  const endDate = new Date(endMs)
  const startDate = new Date(startMs)
  const endUtc = Date.UTC(endDate.getUTCFullYear(), endDate.getUTCMonth(), endDate.getUTCDate())
  let startUtc = Date.UTC(
    startDate.getUTCFullYear(),
    startDate.getUTCMonth(),
    startDate.getUTCDate(),
  )
  const dayMs = 86_400_000
  const diffDays = Math.floor((endUtc - startUtc) / dayMs)
  if (diffDays > maxDays) {
    startUtc = endUtc - (maxDays - 1) * dayMs
  }
  const dates: string[] = []
  let cursor = endUtc
  while (cursor >= startUtc) {
    const d = new Date(cursor)
    const yyyy = d.getUTCFullYear().toString().padStart(4, '0')
    const mm = (d.getUTCMonth() + 1).toString().padStart(2, '0')
    const dd = d.getUTCDate().toString().padStart(2, '0')
    dates.push(`${yyyy}-${mm}-${dd}`)
    cursor -= dayMs
  }
  return dates
}

function enoent(path: string): Error {
  const e = new Error(`ENOENT: ${path}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

interface PathParts {
  path: PathSpec
  prefix: string
  raw: string
  key: string
  virtualKey: string
}

function normalizePath(path: PathSpec): PathParts {
  const prefix = path.prefix
  let raw = path.pattern !== null ? path.directory : path.original
  if (prefix !== '' && raw.startsWith(prefix)) {
    raw = raw.slice(prefix.length) || '/'
  }
  const key = stripSlash(raw)
  const virtualKey = key !== '' ? `${prefix}/${key}` : prefix !== '' ? prefix : '/'
  return { path, prefix, raw, key, virtualKey }
}

function readdirRoot(prefix: string): string[] {
  return [`${prefix}/channels`, `${prefix}/dms`, `${prefix}/users`]
}

async function readdirChannels(
  accessor: SlackAccessor,
  prefix: string,
  virtualKey: string,
  index: IndexCacheStore | undefined,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) {
      return listing.entries
    }
  }
  const channels = await listChannels(accessor)
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const ch of channels) {
    const entry = SlackIndexEntry.channel(ch)
    entries.push([entry.vfsName, entry])
    names.push(`${prefix}/channels/${entry.vfsName}`)
  }
  if (index !== undefined) {
    await index.setDir(virtualKey, entries)
  }
  return names
}

async function readdirDms(
  accessor: SlackAccessor,
  prefix: string,
  virtualKey: string,
  index: IndexCacheStore | undefined,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) {
      return listing.entries
    }
  }
  const dms = await listDms(accessor)
  const users = await listUsers(accessor)
  const userMap: Record<string, string> = {}
  for (const u of users) userMap[u.id] = u.name ?? u.id
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const dm of dms) {
    const entry = SlackIndexEntry.dm(dm, userMap)
    entries.push([entry.vfsName, entry])
    names.push(`${prefix}/dms/${entry.vfsName}`)
  }
  if (index !== undefined) {
    await index.setDir(virtualKey, entries)
  }
  return names
}

async function readdirUsers(
  accessor: SlackAccessor,
  prefix: string,
  virtualKey: string,
  index: IndexCacheStore | undefined,
): Promise<string[]> {
  if (index !== undefined) {
    const listing = await index.listDir(virtualKey)
    if (listing.entries !== undefined && listing.entries !== null) {
      return listing.entries
    }
  }
  const users = await listUsers(accessor)
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const u of users) {
    const entry = SlackIndexEntry.user(u)
    entries.push([entry.vfsName, entry])
    names.push(`${prefix}/users/${entry.vfsName}`)
  }
  if (index !== undefined) {
    await index.setDir(virtualKey, entries)
  }
  return names
}

async function readdirChannelDates(
  accessor: SlackAccessor,
  parts: PathParts,
  container: 'channels' | 'dms',
  index: IndexCacheStore | undefined,
): Promise<string[]> {
  if (index === undefined) {
    throw enoent(parts.raw)
  }
  let lookup = await index.get(parts.virtualKey)
  if (lookup.entry === undefined || lookup.entry === null) {
    const parentPath = `${parts.prefix}/${container}`
    const parent = new PathSpec({
      original: parentPath,
      directory: parentPath,
      prefix: parts.prefix,
    })
    await readdir(accessor, parent, index)
    lookup = await index.get(parts.virtualKey)
  }
  if (lookup.entry === undefined || lookup.entry === null) {
    throw enoent(parts.raw)
  }
  const listing = await index.listDir(parts.virtualKey)
  if (listing.entries !== undefined && listing.entries !== null) {
    return listing.entries
  }
  const created = Number.parseInt(lookup.entry.remoteTime || '0', 10) || 0
  const latestTs = await latestMessageTs(accessor, lookup.entry.id)
  let dates: string[]
  if (latestTs !== null && latestTs !== 0 && created !== 0) {
    dates = dateRange(latestTs, created)
  } else if (latestTs !== null && latestTs !== 0) {
    dates = dateRange(latestTs, Math.floor(latestTs))
  } else {
    dates = []
  }
  const entries: [string, IndexEntry][] = []
  const names: string[] = []
  for (const d of dates) {
    const entry = SlackIndexEntry.dateDir(lookup.entry.id, d)
    entries.push([entry.vfsName, entry])
    names.push(`${parts.prefix}/${parts.key}/${entry.vfsName}`)
  }
  await index.setDir(parts.virtualKey, entries)
  return names
}

async function fetchDay(
  accessor: SlackAccessor,
  channelId: string,
  dateStr: string,
  dateVirtualKey: string,
  index: IndexCacheStore,
): Promise<void> {
  let messages: SlackMessage[]
  try {
    messages = await fetchMessagesForDay(accessor, channelId, dateStr)
  } catch (err) {
    if (isSoftHistoryError(err)) {
      await index.setDir(dateVirtualKey, [])
      return
    }
    throw err
  }
  const chatEntry = SlackIndexEntry.chatJsonl(channelId, dateStr)
  const filesEntry = SlackIndexEntry.filesDir(channelId, dateStr)
  await index.setDir(dateVirtualKey, [
    ['chat.jsonl', chatEntry],
    ['files', filesEntry],
  ])
  const fileEntries: [string, IndexEntry][] = []
  for (const msg of messages) {
    const files = (msg.files as { id?: string }[] | undefined) ?? []
    for (const fmeta of files) {
      if (fmeta.id === undefined || fmeta.id === '') continue
      const entry = SlackIndexEntry.file(
        fmeta as Parameters<typeof SlackIndexEntry.file>[0],
        channelId,
        dateStr,
        typeof msg.ts === 'string' ? msg.ts : '',
      )
      fileEntries.push([entry.vfsName, entry])
    }
  }
  await index.setDir(`${dateVirtualKey}/files`, fileEntries)
}

async function readdirDateContents(
  accessor: SlackAccessor,
  parts: PathParts,
  container: 'channels' | 'dms',
  segments: string[],
  index: IndexCacheStore | undefined,
): Promise<string[]> {
  if (index === undefined) throw enoent(parts.raw)
  const cached = await index.listDir(parts.virtualKey)
  if (cached.entries !== undefined && cached.entries !== null) {
    return cached.entries
  }
  const chanSeg = segments[1]
  const dateStr = segments[2]
  if (chanSeg === undefined || dateStr === undefined) throw enoent(parts.raw)
  const parentVirtual = `${parts.prefix}/${container}/${chanSeg}`
  let parentLookup = await index.get(parentVirtual)
  if (parentLookup.entry === undefined || parentLookup.entry === null) {
    const parent = new PathSpec({
      original: parentVirtual,
      directory: parentVirtual,
      prefix: parts.prefix,
    })
    await readdir(accessor, parent, index)
    parentLookup = await index.get(parentVirtual)
  }
  if (parentLookup.entry === undefined || parentLookup.entry === null) {
    throw enoent(parts.raw)
  }
  await fetchDay(accessor, parentLookup.entry.id, dateStr, parts.virtualKey, index)
  const refreshed = await index.listDir(parts.virtualKey)
  if (refreshed.entries !== undefined && refreshed.entries !== null) {
    return refreshed.entries
  }
  throw enoent(parts.raw)
}

async function readdirFilesDir(
  accessor: SlackAccessor,
  parts: PathParts,
  container: 'channels' | 'dms',
  segments: string[],
  index: IndexCacheStore | undefined,
): Promise<string[]> {
  if (index === undefined) throw enoent(parts.raw)
  const cached = await index.listDir(parts.virtualKey)
  if (cached.entries !== undefined && cached.entries !== null) {
    return cached.entries
  }
  const chanSeg = segments[1]
  const dateStr = segments[2]
  if (chanSeg === undefined || dateStr === undefined) throw enoent(parts.raw)
  const datePath = `${parts.prefix}/${container}/${chanSeg}/${dateStr}`
  const dateSpec = new PathSpec({
    original: datePath,
    directory: datePath,
    prefix: parts.prefix,
  })
  await readdir(accessor, dateSpec, index)
  const refreshed = await index.listDir(parts.virtualKey)
  if (refreshed.entries !== undefined && refreshed.entries !== null) {
    return refreshed.entries
  }
  throw enoent(parts.raw)
}

export async function readdir(
  accessor: SlackAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<string[]> {
  const parts = normalizePath(path)
  const { prefix, key, virtualKey } = parts

  if (key === '') return readdirRoot(prefix)
  if (key === 'channels') return readdirChannels(accessor, prefix, virtualKey, index)
  if (key === 'dms') return readdirDms(accessor, prefix, virtualKey, index)
  if (key === 'users') return readdirUsers(accessor, prefix, virtualKey, index)

  const scope = detectScope(path)
  const container = scope.container
  const segments = key.split('/')
  if (container !== 'channels' && container !== 'dms') return []

  if (segments.length === 2) {
    return readdirChannelDates(accessor, parts, container, index)
  }
  if (scope.target === 'date') {
    return readdirDateContents(accessor, parts, container, segments, index)
  }
  if (scope.target === 'files' && segments.length === 4) {
    return readdirFilesDir(accessor, parts, container, segments, index)
  }
  return []
}

export type { SlackChannel }

File: python/mirage/core/slack/readdir.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import logging
from datetime import datetime, timedelta, timezone

from mirage.accessor.slack import SlackAccessor
from mirage.cache.index import IndexCacheStore, IndexEntry
from mirage.core.slack._client import slack_get
from mirage.core.slack.channels import list_channels, list_dms
from mirage.core.slack.files import file_blob_name
from mirage.core.slack.formatters import (channel_dirname, dm_dirname,
                                          user_filename)
from mirage.core.slack.history import fetch_messages_for_day
from mirage.core.slack.scope import SlackScope, detect_scope
from mirage.core.slack.users import list_users
from mirage.types import PathSpec
from mirage.utils.errors import enoent

logger = logging.getLogger(__name__)

VIRTUAL_ROOTS = ("channels", "dms", "users")

_SOFT_HISTORY_ERRORS = (
    "not_in_channel",
    "channel_not_found",
    "missing_scope",
    "is_archived",
    "not_authed",
)


def _date_range(latest_ts: float,
                created: int,
                max_days: int = 90) -> list[str]:
    end = datetime.fromtimestamp(latest_ts, tz=timezone.utc).date()
    start = datetime.fromtimestamp(created, tz=timezone.utc).date()
    if (end - start).days > max_days:
        start = end - timedelta(days=max_days - 1)
    dates = []
    d = end
    while d >= start:
        dates.append(d.isoformat())
        d -= timedelta(days=1)
    return dates


async def _latest_message_ts(config, channel_id: str) -> float | None:
    try:
        data = await slack_get(config,
                               "conversations.history",
                               params={
                                   "channel": channel_id,
                                   "limit": 1,
                               })
    except RuntimeError as e:
        if any(code in str(e) for code in _SOFT_HISTORY_ERRORS):
            logger.debug(
                "slack: history denied for %s (%s); treating as empty",
                channel_id, e)
            return None
        raise
    messages = data.get("messages", [])
    if messages:
        return float(messages[0].get("ts", "0"))
    return None


def _normalize_path(path: PathSpec | str) -> tuple[PathSpec, str, str, str]:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    prefix = path.prefix or ""
    raw = path.directory if path.pattern else path.original
    if prefix and raw.startswith(prefix):
        raw = raw[len(prefix):] or "/"
    key = raw.strip("/")
    virtual_key = prefix + "/" + key if key else prefix or "/"
    return path, prefix, key, virtual_key


async def _readdir_root(prefix: str) -> list[str]:
    return [f"{prefix}/channels", f"{prefix}/dms", f"{prefix}/users"]


async def _readdir_channels(
    accessor: SlackAccessor,
    prefix: str,
    virtual_key: str,
    index: IndexCacheStore | None,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    channels = await list_channels(accessor.config)
    entries: list[tuple[str, IndexEntry]] = []
    names: list[str] = []
    for ch in channels:
        dirname = channel_dirname(ch)
        entry = IndexEntry(
            id=ch["id"],
            name=ch.get("name", ""),
            resource_type="slack/channel",
            vfs_name=dirname,
            remote_time=str(ch.get("created", 0)),
        )
        entries.append((dirname, entry))
        names.append(f"{prefix}/channels/{dirname}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_dms(
    accessor: SlackAccessor,
    prefix: str,
    virtual_key: str,
    index: IndexCacheStore | None,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    dms = await list_dms(accessor.config)
    users = await list_users(accessor.config)
    user_map = {u["id"]: u.get("name", u["id"]) for u in users}
    entries: list[tuple[str, IndexEntry]] = []
    names: list[str] = []
    for dm in dms:
        dirname = dm_dirname(dm, user_map)
        uid = dm.get("user", "")
        entry = IndexEntry(
            id=dm["id"],
            name=user_map.get(uid, uid),
            resource_type="slack/dm",
            vfs_name=dirname,
            remote_time=str(dm.get("created", 0)),
        )
        entries.append((dirname, entry))
        names.append(f"{prefix}/dms/{dirname}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_users(
    accessor: SlackAccessor,
    prefix: str,
    virtual_key: str,
    index: IndexCacheStore | None,
) -> list[str]:
    if index is not None:
        listing = await index.list_dir(virtual_key)
        if listing.entries is not None:
            return listing.entries
    users = await list_users(accessor.config)
    entries: list[tuple[str, IndexEntry]] = []
    names: list[str] = []
    for u in users:
        filename = user_filename(u)
        entry = IndexEntry(
            id=u["id"],
            name=u.get("name", ""),
            resource_type="slack/user",
            vfs_name=filename,
        )
        entries.append((filename, entry))
        names.append(f"{prefix}/users/{filename}")
    if index is not None:
        await index.set_dir(virtual_key, entries)
    return names


async def _readdir_channel_dates(
    accessor: SlackAccessor,
    path: PathSpec,
    prefix: str,
    key: str,
    virtual_key: str,
    container: str,
    index: IndexCacheStore | None,
) -> list[str]:
    if index is None:
        raise enoent(path)
    lookup = await index.get(virtual_key)
    if lookup.entry is None:
        parent_str = prefix + "/" + container
        parent = PathSpec(original=parent_str,
                          directory=parent_str,
                          prefix=prefix)
        await readdir(accessor, parent, index)
        lookup = await index.get(virtual_key)
    if lookup.entry is None:
        raise enoent(path)
    listing = await index.list_dir(virtual_key)
    if listing.entries is not None:
        return listing.entries
    created = int(lookup.entry.remote_time or 0)
    latest_ts = await _latest_message_ts(accessor.config, lookup.entry.id)
    if latest_ts and created:
        dates = _date_range(latest_ts, created)
    elif latest_ts:
        dates = _date_range(latest_ts, int(latest_ts))
    else:
        dates = []
    entries: list[tuple[str, IndexEntry]] = []
    names: list[str] = []
    for d in dates:
        entry = IndexEntry(
            id=f"{lookup.entry.id}:{d}",
            name=d,
            resource_type="slack/date_dir",
            vfs_name=d,
        )
        entries.append((d, entry))
        names.append(f"{prefix}/{key}/{d}")
    await index.set_dir(virtual_key, entries)
    return names


async def _readdir_date_contents(
    accessor: SlackAccessor,
    path: PathSpec,
    prefix: str,
    key: str,
    virtual_key: str,
    container: str,
    chan_seg: str,
    date_str: str,
    index: IndexCacheStore | None,
) -> list[str]:
    if index is None:
        raise enoent(path)
    cached = await index.list_dir(virtual_key)
    if cached.entries is not None:
        return cached.entries
    parent_vk = f"{prefix}/{container}/{chan_seg}"
    parent_lookup = await index.get(parent_vk)
    if parent_lookup.entry is None:
        parent_str = f"{prefix}/{container}/{chan_seg}"
        parent = PathSpec(original=parent_str,
                          directory=parent_str,
                          prefix=prefix)
        await readdir(accessor, parent, index)
        parent_lookup = await index.get(parent_vk)
    if parent_lookup.entry is None:
        raise enoent(path)
    channel_id = parent_lookup.entry.id
    await _fetch_day(accessor, channel_id, date_str, virtual_key, index)
    cached = await index.list_dir(virtual_key)
    if cached.entries is not None:
        return cached.entries
    raise enoent(path)


async def _readdir_files_dir(
    accessor: SlackAccessor,
    path: PathSpec,
    prefix: str,
    key: str,
    virtual_key: str,
    container: str,
    chan_seg: str,
    date_str: str,
    index: IndexCacheStore | None,
) -> list[str]:
    if index is None:
        raise enoent(path)
    cached = await index.list_dir(virtual_key)
    if cached.entries is not None:
        return cached.entries
    date_str_path = f"{prefix}/{container}/{chan_seg}/{date_str}"
    date_path = PathSpec(original=date_str_path,
                         directory=date_str_path,
                         prefix=prefix)
    await readdir(accessor, date_path, index)
    cached = await index.list_dir(virtual_key)
    if cached.entries is not None:
        return cached.entries
    raise enoent(path)


async def readdir(
    accessor: SlackAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> list[str]:
    path, prefix, key, virtual_key = _normalize_path(path)

    if not key:
        return await _readdir_root(prefix)

    scope = detect_scope(path)
    container = scope.container

    if key == "channels":
        return await _readdir_channels(accessor, prefix, virtual_key, index)
    if key == "dms":
        return await _readdir_dms(accessor, prefix, virtual_key, index)
    if key == "users":
        return await _readdir_users(accessor, prefix, virtual_key, index)

    parts = key.split("/")
    if container in ("channels", "dms") and len(parts) == 2:
        return await _readdir_channel_dates(accessor, path, prefix, key,
                                            virtual_key, container, index)
    if container in ("channels", "dms") and scope.target == "date":
        return await _readdir_date_contents(accessor, path, prefix, key,
                                            virtual_key, container, parts[1],
                                            parts[2], index)
    if container in ("channels", "dms") and scope.target == "files":
        return await _readdir_files_dir(accessor, path, prefix, key,
                                        virtual_key, container, parts[1],
                                        parts[2], index)
    return []


async def _fetch_day(
    accessor: SlackAccessor,
    channel_id: str,
    date_str: str,
    date_vkey: str,
    index: IndexCacheStore,
) -> None:
    try:
        messages = await fetch_messages_for_day(accessor.config, channel_id,
                                                date_str)
    except RuntimeError as e:
        if any(code in str(e) for code in _SOFT_HISTORY_ERRORS):
            logger.debug("slack: history denied for %s/%s (%s); empty day",
                         channel_id, date_str, e)
            await index.set_dir(date_vkey, [])
            return
        raise
    chat_entry = IndexEntry(
        id=f"{channel_id}:{date_str}:chat",
        name="chat.jsonl",
        resource_type="slack/chat_jsonl",
        vfs_name="chat.jsonl",
    )
    files_entry = IndexEntry(
        id=f"{channel_id}:{date_str}:files",
        name="files",
        resource_type="slack/files_dir",
        vfs_name="files",
    )
    await index.set_dir(date_vkey, [
        ("chat.jsonl", chat_entry),
        ("files", files_entry),
    ])
    file_entries: list[tuple[str, IndexEntry]] = []
    for msg in messages:
        for fmeta in msg.get("files", []) or []:
            if not fmeta.get("id"):
                continue
            blob_name = file_blob_name(fmeta)
            file_entries.append(
                (blob_name,
                 IndexEntry(
                     id=fmeta["id"],
                     name=fmeta.get("title") or fmeta.get("name") or "",
                     resource_type="slack/file",
                     vfs_name=blob_name,
                     size=fmeta.get("size"),
                     remote_time=str(fmeta.get("timestamp", "")),
                     extra={
                         "mimetype":
                         fmeta.get("mimetype", ""),
                         "url_private_download":
                         fmeta.get("url_private_download", ""),
                         "filetype":
                         fmeta.get("filetype", ""),
                         "ts":
                         msg.get("ts", ""),
                         "channel_id":
                         channel_id,
                         "date":
                         date_str,
                     },
                 )))
    await index.set_dir(date_vkey + "/files", file_entries)


__all__ = ["readdir", "VIRTUAL_ROOTS", "SlackScope"]
